### PR TITLE
Custom (station) map loading for admin use.

### PR DESCRIPTION
**Pull request by TheChosenEvilOne**
## About The Pull Request

Allows admins to load custom maps for the station.
Sadly this is limited to 1MB because of input being used for the file.

## Why It's Good For The Game

Allows for some fun admin stuff.

## Changelog
:cl:
admin: Added an option to set the next map to a custom map.
/:cl:

### CODE REVIEW

**Comment by MMMiracles** [comment]
I could see this being useful for annual events like the ball or toolbox for a cause, but not for actual map testing since even more basic maps like Box beat the 1MB threshold.

Still pretty neat, from an admeme and debugging perspective.

**Comment by TheChosenEvilOne** [comment]
> 
> 
> I could see this being useful for annual events like the ball or toolbox for a cause, but not for actual map testing since even more basic maps like Box beat the 1MB threshold.
> 
> Still pretty neat, from an admeme and debugging perspective.

Yeah, the upload limit is a bit too low for bigger maps but I do actually have two maps that I want to test that are under 1MB.

**Comment by MCterra10** [comment]
If you wanna test stuff, you can just change the define locally

**Comment by TheChosenEvilOne** [comment]
> 
> 
> If you wanna test stuff, you can just change the define locally

If I wanted to test stuff locally I would not have done this.

**Comment by cacogen** [comment]
can't wait for a high pop byos event

File: code/modules/admin/verbs/maprotation.dm [hunk 1]
```diff
@@ -33,12 +33,69 @@
...
+		
+		if(copytext("[map_file]",-4) != ".dmm")
+			to_chat(src, "<span class='warning'>Filename must end in '.dmm': [map_file]</span>")
+			return
+
+		if(!fcopy(map_file, "_maps/custom/[map_file]"))
+			return
+		
+		// This is to make sure the map works so the server does not start without a map.
+		var/datum/parsed_map/M = new (map_file)
```
Comment (AnturK) [inline code/modules/admin/verbs/maprotation.dm]: You should check if parsed bounds are set, also might want to call the general validation

**Review by AnturK — CHANGES_REQUESTED** [review]

File: code/modules/admin/verbs/maprotation.dm (hunk 1 above)
Comment (TheChosenEvilOne) [reply code/modules/admin/verbs/maprotation.dm]: That should do it.

File: code/controllers/subsystem/mapping.dm [hunk 2]
```diff
@@ -233,7 +233,12 @@ SUBSYSTEM_DEF(mapping)
 	station_start = world.maxz + 1
 	INIT_ANNOUNCE("Loading [config.map_name]...")
 	LoadGroup(FailedZs, "Station", config.map_path, config.map_file, config.traits, ZTRAITS_STATION)
-
+	// Custom maps are removed after station loading so the map files does not persist for no reason.
+	if(config.map_path == "custom")
```
Comment (AnturK) [inline code/controllers/subsystem/mapping.dm]: DB should get the the correct value though, this needs to be after that.

**Review by AnturK — CHANGES_REQUESTED** [review]

**Pull request merged**